### PR TITLE
Fixed: handling Circular children with toJSON overidden (issue #13)

### PR DESCRIPTION
**Pull request by glesperance**
Fix for issue #13

### CODE REVIEW

**Comment by mcollina** [comment]
Thanks for your contribution!

Do you think https://github.com/davidmarkclements/fast-safe-stringify/blob/master/index.js#L4-L6 is still needed?

**Comment by glesperance** [comment]
@mcollina No you're right ; it is not. Just updated the PR.

Also ; this might be a premature optimization but we could also short circuit the whole process whenever an object passes a `toJSON` method or simply isn't an object

such that `function stringify` becomes:
```
function stringify (obj) {
  if (typeof obj === 'object' && typeof obj.toJSON !== 'function') {
    decirc(obj, '', [], null)
  }
  return JSON.stringify(obj)
}
```

**Comment by mcollina** [comment]
@glesperance 👍 , go for it

**Comment by glesperance** [comment]
Updated ; should be ready to merge & release :)

File: test.js [hunk 1]
```diff
@@ -153,3 +153,34 @@ test('double child circular reference', function (assert) {
   assert.deepEqual(fixture, cloned)
   assert.end()
 })
+
+test('child circular reference with toJSON', function (assert) {
+  // Create a test object that has an overriden `toJSON` property
+  TestObject.prototype.toJSON = function () { return { special: 'case' } }
+  function TestObject (content) {}
```
Comment (davidmarkclements) [inline test.js]: no need for content param here

**Comment by davidmarkclements** [comment]
hey really great work @glesperance thanks for the PR

would you mind running benchmarks (`node benchmark`) on the master branch and this branch on your machine and then posting here for comparison?

**Comment by glesperance** [comment]
Here you go

- master:
```
inspectBench*10000: 135.306ms
jsonStringifySafeBench*10000: 32.183ms
fastSafeStringifyBench*10000: 21.717ms
inspectCircBench*10000: 163.313ms
jsonStringifyCircSafeBench*10000: 38.375ms
fastSafeStringifyCircBench*10000: 32.367ms
inspectDeepBench*10000: 1631.096ms
jsonStringifySafeDeepBench*10000: 611.851ms
fastSafeStringifyDeepBench*10000: 283.027ms
inspectDeepCircBench*10000: 1652.218ms
jsonStringifySafeDeepCircBench*10000: 618.628ms
fastSafeStringifyDeepCircBench*10000: 295.040ms
jsonStringifyDeepTryFirstBench*10000: 183.465ms
fastSafeStringifyDeepTryFirstBench*10000: 177.863ms
inspectBench*10000: 101.306ms
jsonStringifySafeBench*10000: 26.375ms
fastSafeStringifyBench*10000: 19.484ms
inspectCircBench*10000: 156.615ms
jsonStringifyCircSafeBench*10000: 39.666ms
fastSafeStringifyCircBench*10000: 30.496ms
inspectDeepBench*10000: 1621.195ms
jsonStringifySafeDeepBench*10000: 649.044ms
fastSafeStringifyDeepBench*10000: 282.504ms
inspectDeepCircBench*10000: 1674.615ms
jsonStringifySafeDeepCircBench*10000: 641.023ms
fastSafeStringifyDeepCircBench*10000: 291.133ms
jsonStringifyDeepTryFirstBench*10000: 194.723ms
fastSafeStringifyDeepTryFirstBench*10000: 186.446ms
```
- child-tojson-fix:
```
inspectBench*10000: 137.680ms
jsonStringifySafeBench*10000: 28.946ms
fastSafeStringifyBench*10000: 22.329ms
inspectCircBench*10000: 163.566ms
jsonStringifyCircSafeBench*10000: 39.524ms
fastSafeStringifyCircBench*10000: 33.226ms
inspectDeepBench*10000: 1725.230ms
jsonStringifySafeDeepBench*10000: 609.582ms
fastSafeStringifyDeepBench*10000: 298.350ms
inspectDeepCircBench*10000: 1753.096ms
jsonStringifySafeDeepCircBench*10000: 671.960ms
fastSafeStringifyDeepCircBench*10000: 316.777ms
jsonStringifyDeepTryFirstBench*10000: 203.530ms
fastSafeStringifyDeepTryFirstBench*10000: 197.191ms
inspectBench*10000: 123.209ms
jsonStringifySafeBench*10000: 27.780ms
fastSafeStringifyBench*10000: 20.665ms
inspectCircBench*10000: 180.578ms
jsonStringifyCircSafeBench*10000: 40.680ms
fastSafeStringifyCircBench*10000: 34.611ms
inspectDeepBench*10000: 1940.454ms
jsonStringifySafeDeepBench*10000: 653.678ms
fastSafeStringifyDeepBench*10000: 306.300ms
inspectDeepCircBench*10000: 1942.596ms
jsonStringifySafeDeepCircBench*10000: 667.188ms
fastSafeStringifyDeepCircBench*10000: 300.606ms
jsonStringifyDeepTryFirstBench*10000: 187.450ms
fastSafeStringifyDeepTryFirstBench*10000: 189.015ms
```

**Comment by davidmarkclements** [comment]
Yah - please!

**Pull request merged**